crates/primitives/src/signature/sig.rs

```diff
@@ -1,821 +1,144 @@
+use crate::{Parity, SignatureError, U256};
 use core::str::FromStr;
 
-/// The order of the secp256k1 curve
-const SECP256K1N_ORDER: U256 =
-    uint!(0xFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFFEBAAEDCE6AF48A03BBFD25E8CD0364141_U256);
-
-/// An Ethereum ECDSA signature.
-#[derive(Clone, Copy, Debug, Hash, PartialEq, Eq)]
-pub struct Signature {
-    v: Parity,
-    r: U256,
-    s: U256,
-}
-
-impl<'a> TryFrom<&'a [u8]> for Signature {
-    type Error = SignatureError;
-
-    /// Parses a raw signature which is expected to be 65 bytes long where
-    /// the first 32 bytes is the `r` value, the second 32 bytes the `s` value
-    /// and the final byte is the `v` value in 'Electrum' notation.
-    fn try_from(bytes: &'a [u8]) -> Result<Self, Self::Error> {
-        if bytes.len() != 65 {
-            return Err(SignatureError::FromBytes("expected exactly 65 bytes"));
-        }
-        Self::from_bytes_and_parity(&bytes[..64], bytes[64] as u64)
-    }
-}
-
-impl FromStr for Signature {
-    type Err = SignatureError;
-
-    fn from_str(s: &str) -> Result<Self, Self::Err> {
-        let bytes = hex::decode(s)?;
-        Self::try_from(&bytes[..])
-    }
-}
-
-impl From<&Signature> for [u8; 65] {
-    #[inline]
-    fn from(value: &Signature) -> [u8; 65] {
-        value.as_bytes()
-    }
-}
-
-impl From<Signature> for [u8; 65] {
-    #[inline]
-    fn from(value: Signature) -> [u8; 65] {
-        value.as_bytes()
-    }
-}
-
-impl From<&Signature> for Vec<u8> {
-    #[inline]
-    fn from(value: &Signature) -> Self {
-        value.as_bytes().to_vec()
-    }
-}
-
-impl From<Signature> for Vec<u8> {
-    #[inline]
-    fn from(value: Signature) -> Self {
-        value.as_bytes().to_vec()
-    }
-}
-
-#[cfg(feature = "k256")]
-impl From<(k256::ecdsa::Signature, k256::ecdsa::RecoveryId)> for Signature {
-    fn from(value: (k256::ecdsa::Signature, k256::ecdsa::RecoveryId)) -> Self {
-        Self::from_signature_and_parity(value.0, value.1).unwrap()
-    }
-}
-
-#[cfg(feature = "k256")]
-impl TryFrom<Signature> for k256::ecdsa::Signature {
-    type Error = k256::ecdsa::Error;
-
-    fn try_from(value: Signature) -> Result<Self, Self::Error> {
-        value.to_k256()
-    }
-}
-
-#[cfg(feature = "rlp")]
-impl Signature {
+/// An Ethereum Generic signature.
+pub trait Signature<'a>:
+    TryFrom<&'a [u8], Error = SignatureError> + FromStr<Err = SignatureError>
+{
     /// Decode an RLP-encoded VRS signature.
-    pub fn decode_rlp_vrs(buf: &mut &[u8]) -> Result<Self, alloy_rlp::Error> {
-        use alloy_rlp::Decodable;
-
-        let parity: Parity = Decodable::decode(buf)?;
-        let r = Decodable::decode(buf)?;
-        let s = Decodable::decode(buf)?;
-
-        Self::from_rs_and_parity(r, s, parity)
-            .map_err(|_| alloy_rlp::Error::Custom("attempted to decode invalid field element"))
-    }
-}
+    #[cfg(feature = "rlp")]
+    fn decode_rlp_vrs(buf: &mut &[u8]) -> Result<Self, alloy_rlp::Error>
+    where
+        Self: Sized;
```

Comment:
this for example can have a default impl, since the trait method `Signature::fn from_rs_and_parity` is accessible